src/module.ts

```diff
@@ -58,12 +58,12 @@ export namespace TypeRef {
   export const F32: TypeRef = 4 /* _BinaryenTypeFloat32 */;
   export const F64: TypeRef = 5 /* _BinaryenTypeFloat64 */;
   export const V128: TypeRef = 6 /* _BinaryenTypeVec128 */;
-  export const Funcref: TypeRef = 7 /* _BinaryenTypeFuncref */;
-  export const Externref: TypeRef = 8 /* _BinaryenTypeExternref */;
-  export const Anyref: TypeRef = 8 /* _BinaryenTypeAnyref */;
-  export const Eqref: TypeRef = 9 /* _BinaryenTypeEqref */;
-  export const I31ref: TypeRef = 10 /* _BinaryenTypeI31ref */;
-  export const Dataref: TypeRef = 11 /* _BinaryenTypeDataref */;
+  export let Funcref: TypeRef = -1 /* ignore */;
+  export let Externref: TypeRef = -1 /* ignore */;
+  export let Anyref: TypeRef = -1 /* ignore */;
+  export let Eqref: TypeRef = -1 /* ignore */;
+  export let I31ref: TypeRef = -1 /* ignore */;
+  export let Dataref: TypeRef = -1 /* ignore */;
```

Comment:
Wondering: Can't these become

```
export const Funcref: TypeRef = _BinaryenTypeFuncref();
```

i.e. always assigning the dynamic value? If `_BinaryenTypeFuncref` is not in a comment, update-constants does not try to substitute the value. Or will that fail because Binaryen isn't fully loaded yet?

Reply:
> Wondering: Can't these become

That was my first attempt. But unfortunately, we should defer these assignments. See this initial commit:
https://github.com/AssemblyScript/assemblyscript/pull/2385/commits/4979a91b5ae9ebf990f53a3aa53141208f6ebee4

Reply:
Reopening as I'm not quite sure how Binaryen isn't fully loaded here. Afaik the changes we've made to Binaryen.js should make sure that the Binaryen instance is fully top-level `await`ed before anything is exported. I might be missing something, though. What's the error here when assigning the values directly?

Reply:
<img width="719" alt="Снимок экрана 2022-07-26 в 14 48 10" src="https://user-images.githubusercontent.com/1301959/180998790-e1eff851-4d7a-41f1-85dd-e4855440b9c6.png">

Reply:
Any idea what `__` is? :)

Reply:
I guess it's mangled by esbuild `_BinaryenTypeFuncref`

Reply:
Yep. If I rearrange inti order I got this:
```
export const Externref: TypeRef = binaryen._BinaryenTypeExternref();
                                             ^

TypeError: f_ is not a function
```

Reply:
From looking at this, it seems that `_BinaryenTypeXY` is not populated yet after an `import "binaryen"`. Seems cursed, currently only works by chance, would assume this is some issue in Binaryen.js.

Reply:
These become populated here

https://github.com/WebAssembly/binaryen/blob/a1464d9e004a43ef6d3f2fa83444e7906d9b620f/src/js/binaryen.js-post.js#L4835-L4843

and perhaps this is a race between `await`ing and `onRuntimeInitialized`. Hmm...

Reply:
But we're using custom binding =) Btw original bindings also defer init constants and wait while `runtimeInitialized` is set to true:
https://github.com/WebAssembly/binaryen/blob/main/src/js/binaryen.js-post.js#L4837

Reply:
In fact we are using Binaryen.js (Wasm variant) underneath, but using our own types which make it look like this was a pure Wasm variant. Quite strange :)

Reply:
In the Wasm variant, the `else` there is taken btw.

Reply:
One thing we are missing is to add `_BinaryenTypeXY` to the exports of `src/glue/binaryen.js`. Then it's not undefined anymore but produces a different error :)

Reply:
> One thing we are missing is to add _BinaryenTypeXY to the exports of src/glue/binaryen.js

Basically I added it:
https://github.com/AssemblyScript/assemblyscript/pull/2385/files#diff-26d003ef41aa7d2210019ca9c4d16999bb863185aeaf9a1fd427019f751e1f1fR26-R31